### PR TITLE
Add response model for `GET /package`

### LINKED ISSUE #200: Create response models for `GET /package` endpoint
Currently the `GET /package` endpoint dumps the entire SQLAlchemy model which contains a lot of unnecessary information, and can change frequently based on the underlying database model. I think it would be better to establish a formal interface that does not necessarily have to match up with the underlying database model 1-to-1.

### PR DESCRIPTION
**Pull request by BruDriguezz**
Closes #200

### CODE REVIEW

**Comment by BruDriguezz** [comment]
@Robin5605 Thoughts?

File: src/mainframe/models/schemas.py [hunk 1]
```diff
@@ -16,6 +16,46 @@ class Error(BaseModel):
...
+            status=scan_result.status,
+            score=scan_result.score,
+            inspector_url=scan_result.inspector_url,
+            rules=[rule.rule_name for rule in scan_result.rules],
+            download_urls=[url.url for url in scan_result.download_urls],
+            queued_at=scan_result.queued_at.isoformat(),
+            queued_by=scan_result.queued_by,
+            pending_at=scan_result.pending_at.isoformat() if scan_result.pending_at else None,
+            pending_by=scan_result.pending_by,
+            finished_at=scan_result.finished_at.isoformat() if scan_result.finished_at else None,
```
Comment (Robin5605) [inline src/mainframe/models/schemas.py:53]: I think these should be UNIX timestamps. The model can be typed as `Optional[int]`, or even better, be typed as `Optional[datetime]` with a custom serializer. The Pydantic docs actually has an exact example for this: https://docs.pydantic.dev/latest/concepts/serialization/#custom-serializers:
```py
class WithCustomEncoders(BaseModel):
    model_config = ConfigDict(ser_json_timedelta='iso8601')

    dt: datetime
    diff: timedelta

    @field_serializer('dt')
    def serialize_dt(self, dt: datetime, _info):
        return dt.timestamp()
```

File: src/mainframe/rules.py [hunk 2]
```diff
@@ -7,7 +7,7 @@
 
 from mainframe.constants import mainframe_settings
 
-REPOSITORY: Final[str] = "vipyrsec/security-intelligence"
+REPOSITORY: Final[str] = "vipyrsec/security-intelligence-public"  # Change back to security-intelligence upon merge
```
Comment (Robin5605) [inline src/mainframe/rules.py:10]:
```suggestion
REPOSITORY: Final[str] = "vipyrsec/security-intelligence"
```

**Review by Robin5605 — CHANGES_REQUESTED** [review]

File: src/mainframe/models/schemas.py [hunk 3]
```diff
@@ -21,40 +24,51 @@ class Package(BaseModel):
...
+            finished_at=from_scan.finished_at,
+            finished_by=from_scan.finished_by,
+            commit_hash=from_scan.commit_hash,
         )
 
+    @field_serializer(
+        "queued_at",
+        "pending_at",
+        "finished_at",
+    )
```
Comment (Robin5605) [inline src/mainframe/models/schemas.py:67]: `reported_at` should also be added here

File: src/mainframe/models/schemas.py [hunk 4]
```diff
@@ -21,40 +24,51 @@ class Package(BaseModel):
...
+    queued_by: str
+    pending_at: Optional[datetime]
+    pending_by: Optional[str]
+    finished_at: Optional[datetime]
+    finished_by: Optional[str]
+    commit_hash: Optional[str]
 
     @classmethod
-    def build_from_db(cls, scan_result): # I'm... not sure of what type `scan_result` is. I'll ask later.
+    def build_from_db(cls, from_scan: Scan):
```
Comment (Robin5605) [inline src/mainframe/models/schemas.py:43]: Would prefer `from_db` as the function name and `scan` as the argument name

**Review by Robin5605 — CHANGES_REQUESTED** [review]

File: src/mainframe/models/schemas.py [hunk 5]
```diff
@@ -16,6 +19,57 @@ class Error(BaseModel):
...
 
 
+class Package(BaseModel):
+    """Model representing a package queried from the database."""
+
+    scan_id: str
+    name: str
+    version: str = ""
+    status: str = ""
+    score: Optional[int] = 0  # Just for the sake of texting, remove upon merge
```
Comment (Robin5605) [inline src/mainframe/models/schemas.py:29]: `score` should always be returned (even if it's 0)
```suggestion
    score: int
```

File: src/mainframe/models/schemas.py [hunk 6]
```diff
@@ -16,6 +19,57 @@ class Error(BaseModel):
...
+    inspector_url: Optional[str] = ""
+    rules: list[str] = []
+    download_urls: list[str] = []
+    queued_at: str = ""
+    queued_by: str = ""
+    pending_at: Optional[str] = None
+    pending_by: Optional[str] = None
+    finished_at: Optional[str] = None
+    finished_by: Optional[str] = None
+    commit_hash: Optional[str] = ""
```
Comment (Robin5605) [inline src/mainframe/models/schemas.py:39]: Appears to be missing `reported_at` and `reported_by` field. I'd like for this information to be available as well so the client can tell if it's been reported already.

File: src/mainframe/models/schemas.py [hunk 7]
```diff
@@ -16,6 +19,57 @@ class Error(BaseModel):
...
+        return cls(
+            scan_id=str(scan.scan_id),
+            name=scan.name,
+            version=scan.version,
+            status=str(scan.status),
+            score=scan.score,
+            inspector_url=scan.inspector_url,
+            rules=[rule.name for rule in scan.rules],
+            download_urls=[url.url for url in scan.download_urls],
+            reported_at=scan.reported_at,
```
Comment (Robin5605) [inline src/mainframe/models/schemas.py:52]: See above. `reported_at` should also be included

**Review by Robin5605 — CHANGES_REQUESTED** [review]

**Comment by AbooMinister25** [comment]
Can I...yeah

**Comment by AbooMinister25** [comment]
Could you also get test coverage back to 100%? We require it via coveragepy.

File: src/mainframe/models/schemas.py [hunk 8]
```diff
@@ -1,6 +1,9 @@
+from datetime import datetime
 from typing import Optional
 
-from pydantic import BaseModel
+from pydantic import BaseModel, field_serializer
+
+from src.mainframe.models.orm import Scan
```
Comment (Robin5605) [inline src/mainframe/models/schemas.py:6]: `src` is not a module that can be imported. This should also fix the docs CI failing
```suggestion
from mainframe.models.orm import Scan
```

File: tests/test_package.py [hunk 9]
```diff
@@ -234,3 +236,40 @@ def test_submit_duplicate_package(
...
+    )
+
+    pkg = (Package.from_db(scan)).model_dump()
+
+    assert pkg.get("queued_at") == int(queued_at.timestamp())
+    assert pkg.get("name") == name
+    assert pkg.get("version") == version
+    assert pkg.get("score") == score
+    assert pkg.get("queued_by") == queued_by
+    assert pkg.get("reported_by") == reported_by
```
Comment (Robin5605) [inline tests/test_package.py:275]: I don't believe this makes a functional difference but you could just use `Package.from_db(scan)` without calling `.model_dump()` on it and accessing the object normally. Might cut down on duplicated code.

**Review by Robin5605 — CHANGES_REQUESTED** [review]

File: tests/test_package.py (hunk 9 above)
Comment (BruDriguezz) [reply tests/test_package.py:275]: It does make a difference, as the fields are only serialized after you dump it as per Pydantic's documentation.
If you don't, `datetime` isn't capable of being asserted as a POSIX timestamp

Comment (Robin5605) [reply tests/test_package.py:275]: ah, gotcha. looks like you're trying to test two things at once, then (that the datetime fields are serialized into a timestamp properly, and also that the general structure of `Package` is what you expect). it might be worth breaking these two tests apart: one test that tests the structure (what you have now, but don't worry about `.model_dump()`, just compare to a datetime object), and another test that is specifically concerned with the 3 (or 4?) datetime models. There, you will want to `.model_dump`, then compare to `int(queued_at.timestamp()` (similar to what you are doing now, but for all the datetime fields)

Comment (BruDriguezz) [reply tests/test_package.py:275]: I see, got it.
Would you be comfortable with transforming this mark;
```py
@pytest.mark.parametrize(
    "name, queued_at, pending_at, finished_at, reported_at",
    [
        ("pyfoo", datetime(2023, 2, 1, 12, 42, 1), datetime(2023, 9, 1, 16, 33, 0), datetime(2024, 1, 3, 9, 56, 46),
         datetime(2022, 12, 31, 23, 56, 0)),
        ("pybar", datetime(2023, 3, 5, 8, 10, 15), datetime(2023, 10, 12, 13, 45, 30), datetime(2024, 2, 15, 18, 20, 5),
         datetime(2022, 11, 20, 5, 30, 0)),
        ("pybaz", datetime(2023, 4, 10, 16, 20, 30), datetime(2023, 11, 20, 21, 15, 45),
         datetime(2024, 3, 25, 6, 40, 20), datetime(2023, 1, 15, 12, 0, 0)),
    ],
)
``` 
Into a `pytest.fixture`? Then we could do; 
```py
@pytest.fixture
def scan_data():
    return [
        {
            "name": "pyfoo",
            "version": "0.12.1",
            "queued_at": datetime(2023, 2, 1, 12, 42, 1),
            "pending_at": datetime(2023, 9, 1, 16, 33, 0),
            "finished_at": datetime(2024, 1, 3, 9, 56, 46),
            "reported_at": datetime(2022, 12, 31, 23, 56, 0),
        },
        {
            "name": "pybar",
            "version": "3.12.2",
            "queued_at": datetime(2023, 3, 5, 8, 10, 15),
            "pending_at": datetime(2023, 10, 12, 13, 45, 30),
            "finished_at": datetime(2024, 2, 15, 18, 20, 5),
            "reported_at": datetime(2022, 11, 20, 5, 30, 0),
        },
        {
            "name": "pybaz",
            "version": "0.24.8",
            "queued_at": datetime(2023, 4, 10, 16, 20, 30),
            "pending_at": datetime(2023, 11, 20, 21, 15, 45),
            "finished_at": datetime(2024, 3, 25, 6, 40, 20),
            "reported_at": datetime(2023, 1, 15, 12, 0, 0),
        },
    ]

# ...

@pytest.mark.parametrize("scan_data", range(3))
def test_datetime_serialization(scan_data):
    ...
```

I find the test data much more readable as a fixture. Thoughts?

Comment (BruDriguezz) [reply tests/test_package.py:275]: Moreover, scan fixtures could provide the advantage of standardization across multiple tests. Would you be open to considering this approach?

Comment (Robin5605) [reply tests/test_package.py:275]: I would rather not parametrize here at all, to be honest. Test parametrization is most valuable when used in situations such as these: https://github.com/vipyrsec/dragonfly-mainframe/blob/08460a2a0de6b13be43f19cd48ef8b8ba64bd70e/tests/test_package.py#L28-L36
Where we can test edge cases.
In this case, there aren't really "edge cases" and we're just testing the same thing but with different values for the datetimes and whatnot. I'd rather just test one thing like so:
```py
def test_package_from_db():
    scan = Scan(
        name="abc",
        version="def",
        queued_at=datetime(2024, 3, 5, 12, 30, 0),
    )

    pkg = Package.from_db(scan)

    assert pkg.name == "abc"
    assert pkg.version == "version"
    assert pkg.queued_at == datetime(2024, 3, 5, 12, 30, 0)
    # maybe test one or two more fields just to be sure but it's not too important
```

And then testing the timestamp conversions in another test:
```py
def test_package_serializes_datetime():
    pkg = Package(..., queued_at=datetime(2024, 3, 5, 12, 30, 0))
    serialized = pkg.model_dump()
    
    assert serialized.get("queued_at") == 1709663400
    assert serialized.get("reported_at") == ...
    ... # you get the idea
````

Overall I don't think parametrization nets us much value here as we'd simply be testing different values of the same thing, and the code itself is really just passing values around and not manipulating it too much.

Comment (BruDriguezz) [reply tests/test_package.py:275]: I see, that's valid thinking. I'll adapt and commit shortly.

File: src/mainframe/models/schemas.py [hunk 10]
```diff
@@ -16,6 +19,60 @@ class Error(BaseModel):
...
 
 
+class Package(BaseModel):
+    """Model representing a package queried from the database."""
+
+    scan_id: str
+    name: str
+    version: str = ""
+    status: str = ""
+    score: Optional[int]
```
Comment (AbooMinister25) [inline src/mainframe/models/schemas.py:29]: Perhaps I'm being dumb, but I don't see why any of the `Optional` fields can't consistently have a `None` default, or the opposite (not having a default whatsoever), since `Scan` provides defaults for all `Optional` fields anyways.

**Review by Robin5605 — CHANGES_REQUESTED** [review]
Looks great - just one thing. It looks like the permissions for `entrypoint.sh` were changed from `644` to `755` for some reason, can you revert this back?

**Comment by Robin5605** [comment]
Oh, can we get CI passing too? Looks like the error is coming from the fact that we have these two conflicting imports:
```py
from letsbuilda.pypi import Package, PyPIServices  # type: ignore
from mainframe.models.schemas import (
    ...,
    Package,
    ...,
)
```
One of these will have to be renamed into something else (using the `from x import y as z` syntax.

**Review by Robin5605 — DISMISSED** [review]
Looks great! Thanks for the PR!

**Review by import-pandas-as-numpy — APPROVED** [review]
LGTM from the security side of the house.

**Review by shenanigansd — APPROVED** [review]

**Review by Robin5605 — APPROVED** [review]
